pkg/bootstrap/platform/gcp.go

```diff
@@ -249,33 +253,65 @@ func waitForMetadataSuppliers(suppliers []metadataSupplier, md map[string]string
 	return &wg
 }
 
+// Zones are in the form <country_code>-<region_suffix>-<zone_suffix>.
+var zoneRE = regexp.MustCompile(`^([^-]+-[^-]+)-[^-]+$`)
+
 // Converts a GCP zone into a region.
 func zoneToRegion(z string) (string, error) {
-	// Zones are in the form <region>-<zone_suffix>, so capture everything but the suffix.
-	re := regexp.MustCompile("(.*)-.*")
-	m := re.FindStringSubmatch(z)
+	m := zoneRE.FindStringSubmatch(z)
 	if len(m) != 2 {
 		return "", fmt.Errorf("unable to extract region from GCP zone: %s", z)
 	}
 	return m[1], nil
 }
 
+// Returns the zone where the pod is located in.
+func (e *gcpEnv) getPodZone() (string, error) {
+	z := e.cachedZone.Load()
+	if z != "" {
+		return z, nil
+	}
+	ctx, cfn := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cfn()
+	z, err := zoneFn(ctx)
+	if err != nil {
+		log.Warnf("Error fetching GCP zone from the metadata server: %v", err)
+		return "", err
+	}
+	e.cachedZone.Store(z)
+	return z, nil
+}
+
+// Return the zone guessed from the cluster location.
+func zoneFromClusterLocation(loc string) string {
+	if zoneRE.MatchString(loc) {
+		return loc
+	}
+	// Fallback to "-unknown" zone suffix if the given location is regional.
+	return loc + "-unknown"
+}
+
 // Locality returns the GCP-specific region and zone.
 func (e *gcpEnv) Locality() *core.Locality {
 	var l core.Locality
-	loc := e.Metadata()[GCPLocation]
-	if loc == "" {
-		log.Warnf("Error fetching GCP zone: %v", loc)
-		return &l
+	z, err := e.getPodZone()
+	if err != nil {
```

Comment:
When does it make sense to fall back to cluster location? That's just going to confuse the systems downstream.

Reply:
Fixed to not use the cluster location as a fallback. Now the location will be empty if MDS call fails.